### PR TITLE
(feat) Support i18n for forms entry

### PR DESCRIPTION
**Pull request by ibacher**
## Requirements

- [x] This PR has a title that briefly describes the work done, including the ticket number if there is a ticket.
- [x] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide).
- [x] My work includes tests or is validated by existing tests.

## Summary

My re-worked version of the translation loading code. This loads translations at the same time we load the forms themselves; It should support translating nested sub-forms as well, though this is untested.

NB This work will require a forth-coming PR to ngx-formentry.

## Screenshots

Translated Form

### CODE REVIEW

**Comment by ibacher** [comment]
@icrc-jofrancisco @denniskigen @enyachoke If any of you have any insight into the issue outlined in the ticket (how do I get the same TranslateModule injected into both the patient-chart app and the form-entry module), I'd be appreciative.

**Comment by hadijahkyampeire** [comment]
Thanks @ibacher for the rework, I will try to follow up closely to learn from your implementation.

File: packages/esm-form-entry-app/src/app/form-schema/form-schema.service.ts [hunk 1]
```diff
@@ -139,28 +172,62 @@ export class FormSchemaService {
...
+        (formMetadataObject: FormMetadataObject) => {
           if (formMetadataObject.resources.length > 0) {
             const { resources } = formMetadataObject;
-            const valueReferenceUuid = resources?.find(({ name }) => name === 'JSON schema').valueReference;
-            if (valueReferenceUuid) {
-              this.formsResourceService.getFormClobDataByUuid(valueReferenceUuid).subscribe(
-                (clobData: any) => {
-                  observer.next(clobData);
+            const schemaUuid = resources?.find(({ name }) => name === 'JSON schema')?.valueReference;
+            const translationsUuid = resources?.find(({ name }) =>
```
Comment (ibacher) [inline packages/esm-form-entry-app/src/app/form-schema/form-schema.service.ts:185]: We use the naming convention from Iniz to identify the right translations.

File: packages/esm-form-entry-app/src/app/form-schema/form-schema.service.ts [hunk 2]
```diff
@@ -1,59 +1,92 @@
...
-    const cachedCompiledSchema = this.getCachedCompiledSchemaByUuid(formUuid);
-    if (cachedCompiledSchema && cached) {
-      return of(cachedCompiledSchema);
+  public getFormSchemaByUuid(
+    formUuid: string,
+    language: string = 'en',
+    cached: boolean = true,
+  ): Observable<FormSchema> {
+    if (cached) {
+      const cachedCompiledSchema = this.getCachedCompiledSchemaByUuid(formUuid, language);
```
Comment (ibacher) [inline packages/esm-form-entry-app/src/app/form-schema/form-schema.service.ts:27]: Caching now depends on language, not just the form UUID. For cached forms, the translations are pre-loaded into the cached schema. I haven't tested whether or not this works offline properly, though.

File: packages/esm-form-entry-app/src/app/form-schema/form-schema.service.ts [hunk 3]
```diff
@@ -139,28 +172,62 @@ export class FormSchemaService {
...
+                  console.error(`Error while loading translations for ${formUuid} in ${language}`, err);
+                  formTranslations.error(err);
                 },
               );
             } else {
-              observer.next([]);
+              formTranslations.next({});
             }
+
+            zip(from(formSchema), from(formTranslations))
```
Comment (ibacher) [inline packages/esm-form-entry-app/src/app/form-schema/form-schema.service.ts:219]: Basically, we pair each form schema with a dictionary of translations and return that as an object. I'm not sure it's important, but I tried to ensure there's always a value.

**Review by denniskigen — APPROVED** [review]
Thanks, @ibacher!

File: packages/esm-form-entry-app/package.json [hunk 4]
```diff
@@ -89,6 +90,7 @@
     "karma-jasmine-html-reporter": "^1.5.0",
     "openmrs": "next",
     "protractor": "~7.0.0",
+    "rxjs": "6",
```
[the remark this answers is not in the document]
Comment (ibacher) [reply packages/esm-form-entry-app/package.json:93]: I'm happy to adjust it, but "6" is equivalent to "^6.0.0" and "6.x" 😁.

**Pull request merged**